### ISSUE TITLE
Missing helper: "inline" when using a different template directory

### ISSUE THREAD

**Issue by scipilot** [fmvilas/openapi3-generator#6]
Hi, thanks for making this generator. It's such a great idea to be able to generate from your own templates. I intend to make a "partial" application template which fits into an existing server application, so that I can re-generate (just controllers and a sub-router) freely without having to edit any boilerplate to integrate each version.

I have copied your example "express" template into my application (without making any changes yet), but when I use the "t" parameter to use this copy I get a missing handlebars helper error. 

It begins the generation process, makes the folder structure and the readme with my API name, and the editor files, but nothing else. So it's reading my yaml, and finding the templates OK.

If I remove the -t param, it generates from your internal template OK. All I'm trying to do is copy the template folder to another location, and then providing the -t to that new location.

My structure is:

express/src/OAI/generator/templates/express - a copy of your "express" template
express/lib/API/ - output area 

```
 og -o express/lib/API -t express/src/OAI/generator/templates/ myapi.yaml express
Aaww 💩. Something went wrong:
Error: Missing helper: "inline"
    at Object.<anonymous> (/Users/me/.nvm/versions/node/v9.4.0/lib/node_modules/openapi3-generator/node_modules/handlebars/dist/cjs/handlebars/helpers/helper-missing.js:19:13)
    at Object.eval [as main] (eval at createFunctionContext (/Users/me/.nvm/versions/node/v9.4.0/lib/node_modules/openapi3-generator/node_modules/handlebars/dist/cjs/handlebars/compiler/javascript-compiler.js:254:23), <anonymous>:8:85)
    at main (/Users/me/.nvm/versions/node/v9.4.0/lib/node_modules/openapi3-generator/node_modules/handlebars/dist/cjs/handlebars/runtime.js:175:32)
    at ret (/Users/me/.nvm/versions/node/v9.4.0/lib/node_modules/openapi3-generator/node_modules/handlebars/dist/cjs/handlebars/runtime.js:178:12)
    at ret (/Users/me/.nvm/versions/node/v9.4.0/lib/node_modules/openapi3-generator/node_modules/handlebars/dist/cjs/handlebars/compiler/compiler.js:526:21)
    at fs.readFile (/Users/me/.nvm/versions/node/v9.4.0/lib/node_modules/openapi3-generator/lib/generator.js:43:30)
    at FSReqWrap.readFileAfterClose [as oncomplete] (fs.js:528:3)
```

I can see the helper is in the templates/express/.helpers/all.js file. This is being found OK and loaded by ```registerHelpers()``` in ```generator.js```.

I added a log to ```all.js```:
```
console.log("registering inline helper...");
Handlebars.registerHelper('inline', (str) => {
  return str ? str.replace(/\n/g, '') : '';
});
```
which shows up:
```
og -o express/lib/API -t express/src/OAI/generator/templates/ myapi.yaml express
registering inline helper...
Aaww 💩. Something went wrong:
Error: Missing helper: "inline"
```

 Any help would be appreciated.

**Comment by scipilot** [comment]
I've figured out the problem, but I don't know why it's only happening when I move the templates.

It seems the helpers are being attached to a _different instance of Handlebars_ - the one required inside the ```all.js``` which is not the same as the one in the ```generator.js```.

I tried scoping the all.js module in a closure with Handlebars injected via the require - and it then the code generation works OK.

````
all.js

//const Handlebars = require('handlebars');
module.exports = (Handlebars)=>{
...

	Handlebars.registerHelper('inline', (str) =>{
		return str ? str.replace(/\n/g, '') : '';
	});

};
````

and pass this in from the generator:

```
const registerHelpers = config => new Promise((resolve, reject) => {
 ...

  walker.on('file', async (root, stats, next) => {
    try {
      const file_path = path.resolve(config.templates, path.resolve(root, stats.name));
      require(file_path)(Handlebars);
      next();
    } catch (e) {
...
```

So I guess it's some quirk of require's module caching?

**Comment by fmvilas** [comment]
Hi @scipilot. It definitely looks like require's caching messing things up. Will have a look at it as soon as I have time, but if you could take the lead here and fix it, I'd be very grateful.

**Comment by scipilot** [comment]
@fmvilas 
I have submitted a PR with the proposed solution using module constructors to provide DI. I made it backwards-compatible with the old templates, to help migration. 

I have done a fair amount of research and now understand require's vagaries better. The generator only works correctly when using its own templates, or when it's installed in a project with the same Lodash and Handlebars versions as itself.

There are all these template usage scenarios:

1) Using openapi3-generator installed globally 
 a) Using the original templates - require loads the Handlebars from the global project - OK 
 b) Using templates in a plain directory - require "Error: Cannot find module 'lodash'"
 c) Using templates in another Node project
  i) If the project has lodash, handlebars - require loads the local modules, so helpers aren't installed into the generator's handlebars.
  ii) If not, it's the same as 1(b)
  iii) If it has the wrong version, random stuff will happen compiling the templates
2) With openapi3-generator installed locally as a dependency of a project
 a) If the project uses the same version of Handlebars - require loads the same version - OK
 b) If not - require loads the project's version of Handlebars, giving random results (e.g. parse errors, same as 1(c)(iii) )


To reproduce 2(b) or 1(c)(iii)
```
  npm install -g openapi3-generator
  mkdir project1; cd project1
  npm init
  npm install -s handlebars@v3
  npm install -s openapi3-generator
  cp -r node_modules/openapi3-generator/templates/ mytemplates

  node_modules/.bin/og -o API -t mytemplates /path/to/myapi.yaml express
or
  og -o API -t mytemplates /path/to/myapi.yaml express
Aaww 💩. Something went wrong:
Error: Lexical error on line 1. Unrecognized text.
```

I presume this is due to Handlebars 3 in the project not being able to read the Handlebars 4 templates.

To reproduce 1(b), continue from the above.
```   
  cd ..; mkdir test2; cd test2
  cp -r ../project1/node_modules/openapi3-generator mytemplates
  og -o API -t mytemplates/ /path/to/myapi.yaml express

Aaww 💩. Something went wrong:
Error: Cannot find module 'lodash'
```


To reproduce 1(c)(i)
```
 cd ..; mkdir project2; cd project2
  npm init
  npm install -s handlebars
  cp -r ../project1/node_modules/openapi3-generator/templates/ mytemplates
  cp -r ../project1/node_modules/openapi3-generator/templates/ mytemplates
  og -o API -t mytemplates ../myapi.yaml express
Aaww 💩. Something went wrong:
Error: Missing helper: "inline"
```

This is because the require('handlebars') inside templates/express/.helpers/all.js loads a new instance of Handlebars from the project. It attaches the helpers to this, but doesn't return that instance to generator which continues to use its own version, from the global dependencies.

The proposed solution is to dependency-inject the required modules into the template helper registration module:

In registerHelpers() in generator.js:

```      
require(file_path)(_, Handlebars);
```

To allow a more graceful migration for existing template users, this makes it optional:

 ```
     const mod = require(file_path);
      if(typeof mod == "function") mod(_, Handlebars);
```


In all templates (express, markdown) .helpers/*

```
module.exports = (_, Handlebars) =>{
...
}
```

So I'm not sure if you are a fan of this kind of DI, as I can't see it anywhere else in this project.
It does add a structural requirement to the templates, so I have added comments in the template examples explaining why it's required.

thanks
